web/packages/teleport/src/DesktopSession/KeyboardHandler.tsx

```diff
@@ -0,0 +1,271 @@
+/**
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { Platform, getPlatform } from 'design/platform';
+
+import { TdpClient, ButtonState } from 'teleport/lib/tdp';
+import { SyncKeys } from 'teleport/lib/tdp/codec';
+
+/**
+ * Handles keyboard events.
+ */
+export class KeyboardHandler {
+  /**
+   * Cache for keys whose down event has been withheld until the next keydown or keyup,
+   * or to never be sent if this cache is cleared onfocusout.
+   */
+  private withheldDown: Map<string, boolean> = new Map();
```

Comment:
Should these be [unordered] maps? Would some sort of queue structure make more sense?

Reply:
An unordered map indexed by key makes sense because there should never be more than 1 down or up "pending" for any given key.

Consider the state for a single key (say `MetaLeft`). We always start with a DOWN being withheld, which is sent [to te server] either by another key being pressed, or a `MetaLeft` UP being registered. Then, if we get an UP, the UP is delayed on a timeout, and it's sent by either the timeout timing out, another key being pressed (note: I think my current logic misses this, but this is how it should work for correctness), or a `MetaLeft` DOWN being registered. And so on.

IOW, we should never have anything but a single, withheld-with-optional-timeout, per key. That suggests that this should actually just be one Map-like structure that automagically works this way.

Reply:
I ended up going with a queue like you suggested, which better represents the inherently ordered nature of keystokes: 44595889eba30982176bf283d415879cf35e169d